File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/http/ShortcutRequest.java
package ch.rmy.android.http_shortcuts.http;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;

import org.apache.http.HttpHeaders;
import org.jdeferred.Deferred;
import org.jdeferred.Promise;
import org.jdeferred.impl.DeferredObject;

import java.util.HashMap;
import java.util.Map;

import ch.rmy.android.http_shortcuts.realm.models.Shortcut;
import okhttp3.Credentials;

class ShortcutRequest extends Request<ShortcutResponse> {

    private final Deferred<ShortcutResponse, VolleyError, Void> deferred;
    private final Map<String, String> parameters = new HashMap<>();
    private final Map<String, String> headers = new HashMap<>();
    private String bodyContent;
    private String contentType;

    private ShortcutRequest(int method, String url, final Deferred<ShortcutResponse, VolleyError, Void> deferred) {
        super(method, url, new com.android.volley.Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                deferred.reject(error);
            }
        });
        this.deferred = deferred;
        headers.put(HttpHeaders.CONNECTION, "close");
    }

    @Override
    public byte[] getBody() throws AuthFailureError {
        if (bodyContent.isEmpty()) {
            return super.getBody();
        } else {
            byte[] regularBody = super.getBody();
            byte[] customBody = bodyContent.getBytes();
            if (regularBody == null) {
                return customBody;
            }
            byte[] mergedBody = new byte[regularBody.length + customBody.length];

            System.arraycopy(regularBody, 0, mergedBody, 0, regularBody.length);
            System.arraycopy(customBody, 0, mergedBody, regularBody.length, customBody.length);

            return mergedBody;
        }
    }

    @Override
    public Map<String, String> getHeaders() throws AuthFailureError {
        return headers;
    }

    @Override
    public String getBodyContentType() {
        if (contentType != null) {
            return contentType;
        }
        return super.getBodyContentType();
    }

    @Override
    public Map<String, String> getParams() {
        return parameters;
    }

    @Override
    protected Response<ShortcutResponse> parseNetworkResponse(NetworkResponse response) {
        ShortcutResponse shortcutResponse = new ShortcutResponse(response.statusCode, response.headers, response.data);
        return Response.success(shortcutResponse, null);
    }

    @Override
    protected void deliverResponse(ShortcutResponse response) {
        deferred.resolve(response);
    }

    Promise<ShortcutResponse, VolleyError, Void> getPromise() {
        return deferred.promise();
    }

    static class Builder {

        private final ShortcutRequest request;

        Builder(String method, String url) {
            request = new ShortcutRequest(getMethod(method), url, new DeferredObject<ShortcutResponse, VolleyError, Void>());
        }

        private static int getMethod(String method) {
            switch (method) {
                case Shortcut.METHOD_POST:
                    return Request.Method.POST;
                case Shortcut.METHOD_PUT:
                    return Request.Method.PUT;
                case Shortcut.METHOD_DELETE:
                    return Request.Method.DELETE;
                case Shortcut.METHOD_PATCH:
                    return Request.Method.PATCH;
                default:
                    return Request.Method.GET;
            }
        }

        Builder basicAuth(String username, String password) {
            request.headers.put(HttpHeaders.AUTHORIZATION, Credentials.basic(username, password));
            return this;
        }

        Builder body(String body) {
            request.bodyContent = body;
            return this;
        }

        Builder parameter(String key, String value) {
            request.parameters.put(key, value);
            return this;
        }

        Builder header(String key, String value) {
            if (key.equalsIgnoreCase(HttpHeaders.CONTENT_TYPE)) {
                request.contentType = value;
            } else {
                request.headers.put(key, value);
            }
            return this;
        }

        Builder timeout(int timeout) {
            request.setRetryPolicy(new DefaultRetryPolicy(timeout, 0, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
            return this;
        }

        ShortcutRequest build() {
            return request;
        }

    }

}

File: HTTPShortcuts/app/src/main/kotlin/ch/rmy/android/http_shortcuts/http/ShortcutRequest.kt
package ch.rmy.android.http_shortcuts.http

import ch.rmy.android.http_shortcuts.data.models.Shortcut
import ch.rmy.android.http_shortcuts.utils.UserAgentUtil
import com.android.volley.AuthFailureError
import com.android.volley.DefaultRetryPolicy
import com.android.volley.NetworkResponse
import com.android.volley.Request
import com.android.volley.Response
import io.reactivex.SingleEmitter
import okhttp3.Credentials
import org.apache.http.HttpHeaders

internal class ShortcutRequest private constructor(
    method: Int,
    url: String,
    private val emitter: SingleEmitter<ShortcutResponse>
) : Request<ShortcutResponse>(method, url, Response.ErrorListener(emitter::onError)) {

    private val parameters = mutableMapOf<String, String>()
    private val headers = mutableMapOf<String, String>()
    private var bodyContent: String? = null
    private var contentType: String? = null

    init {
        headers[HttpHeaders.CONNECTION] = "close"
        headers[HttpHeaders.USER_AGENT] = UserAgentUtil.userAgent
    }

    @Throws(AuthFailureError::class)
    override fun getBody(): ByteArray? = when {
        contentType?.startsWith("multipart/form-data") == true -> constructFormDataBody().toByteArray()
        contentType?.startsWith("application/x-www-form-urlencoded") == true -> super.getBody() ?: ByteArray(0)
        else -> (bodyContent ?: "").toByteArray()
    }

    @Throws(AuthFailureError::class)
    override fun getHeaders(): Map<String, String> = headers

    override fun getBodyContentType(): String = contentType ?: DEFAULT_CONTENT_TYPE

    public override fun getParams() = parameters

    override fun parseNetworkResponse(response: NetworkResponse): Response<ShortcutResponse> =
        Response.success(ShortcutResponse(url, response.headers, response.statusCode, response.data), null)

    override fun deliverResponse(response: ShortcutResponse) {
        emitter.onSuccess(response)
    }

    private fun constructFormDataBody(): String =
        StringBuilder("\r\n")
            .apply {
                parameters.entries.forEach { entry ->
                    append("\r\n--$FORM_MULTIPART_BOUNDARY\n")
                    append("Content-Disposition: form-data; name=\"${entry.key}\"")
                    append("\r\n\r\n")
                    append(entry.value)
                }
                append("\n--$FORM_MULTIPART_BOUNDARY--\n")
            }.toString()

    internal class Builder(method: String, url: String, emitter: SingleEmitter<ShortcutResponse>) {

        private val request: ShortcutRequest

        init {
            request = ShortcutRequest(getMethod(method), url, emitter)
        }

        private fun getMethod(method: String) = when (method) {
            Shortcut.METHOD_POST -> Method.POST
            Shortcut.METHOD_PUT -> Method.PUT
            Shortcut.METHOD_DELETE -> Method.DELETE
            Shortcut.METHOD_PATCH -> Method.PATCH
            Shortcut.METHOD_OPTIONS -> Method.OPTIONS
            Shortcut.METHOD_HEAD -> Method.HEAD
            Shortcut.METHOD_TRACE -> Method.TRACE
            else -> Method.GET
        }

        fun contentType(contentType: String) = also {
            request.contentType = contentType
        }

        fun basicAuth(username: String, password: String) = also {
            request.headers[HttpHeaders.AUTHORIZATION] = Credentials.basic(username, password)
        }

        fun body(body: String) = also {
            request.bodyContent = body
        }

        fun parameter(key: String, value: String) = also {
            request.parameters[key] = value
        }

        fun header(key: String, value: String) = also {
            if (key.equals(HttpHeaders.CONTENT_TYPE, ignoreCase = true)) {
                request.contentType = value
            }
            request.headers[key] = value
        }

        fun timeout(timeout: Int) = also {
            request.retryPolicy = DefaultRetryPolicy(timeout, 0, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        }

        fun build() = request

    }

    companion object {

        const val FORM_MULTIPART_BOUNDARY = "----53014704754052338"
        private const val DEFAULT_CONTENT_TYPE = "text/plain"

    }

}
